src/pydrex/diagnostics.py

```diff
@@ -0,0 +1,190 @@
+import itertools as it
+
+import numpy as np
+from numpy import random as rn
+import scipy.linalg as la
+
+import pydrex.tensors as _tensors
+
+
+def Bingham_average(orientations, axis="a"):
+    """Compute Bingham averages from olivine orientation matrices.
+
+    Returns the antipodally symmetric average orientation
+    of the given crystallographic `axis`, or the a-axis by default.
+    Valid axis specifiers are "a" for [100], "b" for [010] and "c" for [001].
+
+    See e.g. [Watson 1966].
+
+    [Watson 1966]: https://doi.org/10.1086%2F627211
+
+    """
+    if axis == "a":
+        row = 0
+    elif axis == "b":
+        row = 1
+    elif axis == "c":
+        row = 2
+    else:
+        raise ValueError(f"axis must be 'a', 'b', or 'c', not {axis}")
+
+    # Scatter (inertia) matrix, see eq. 2.4 in Watson 1966.
+    scatter = np.zeros((3, 3))
+    scatter[0, 0] = np.sum(orientations[:, row, 0] ** 2)
+    scatter[1, 1] = np.sum(orientations[:, row, 1] ** 2)
+    scatter[2, 2] = np.sum(orientations[:, row, 2] ** 2)
+    scatter[1, 0] = np.sum(orientations[:, row, 0] * orientations[:, row, 1])
+    scatter[2, 0] = np.sum(orientations[:, row, 0] * orientations[:, row, 2])
+    scatter[2, 1] = np.sum(orientations[:, row, 1] * orientations[:, row, 2])
+
+    # https://courses.eas.ualberta.ca/eas421/lecturepages/orientation.html
+    # Eigenvector corresponding to largest eigenvalue is the mean direction.
+    # SciPy returns eigenvalues in ascending order (same order for vectors).
+    mean_vector = la.eigh(scatter)[1][:, -1]
+    return mean_vector / la.norm(mean_vector)
+
+
+def resample_orientations(orientations, fractions, n_samples=None):
+    """Generate new samples from `orientations` based on volume distribution.
+
+    If the optional number of samples is not specified,
+    it will be set to the number of original "grains" (length of `fractions`).
+
+    """
+    if n_samples is None:
+        n_samples = len(fractions)
+    # sort_ascending = np.argsort(fractions)[::-1]
+    # frac_ascending = np.take(fractions, sort_ascending)
+    # Cumulative volume fractions.
+    cumfrac = fractions.sort().cumsum()
+    # Number of new samples with volume less than each cumulative fraction.
+    rng = rn.default_rng()
+    count_less = np.searchsorted(cumfrac, rng.random(n_samples))
+    return orientations[count_less]
```

Comment:
Hmm, wait here. Based on what you explained above, should not we actually bring back `argsort` and then write the following:
```
def resample_orientations(orientations, fractions, n_samples=None):
    """Generate new samples from `orientations` based on volume distribution.

    If the optional number of samples is not specified,
    it will be set to the number of original "grains" (length of `fractions`).
    """
    if n_samples is None:
        n_samples = len(fractions)
    sort_ascending = np.argsort(fractions)
    # Cumulative volume fractions.
    cumfrac = fractions[sort_ascending].cumsum()
    # Number of new samples with volume less than each cumulative fraction.
    rng = rn.default_rng()
    count_less = np.searchsorted(cumfrac, rng.random(n_samples))
    return orientations[sort_ascending][count_less]
```
? Also, don't you also need to return the associated volume fractions? It is mentioned in your explanation.

Reply:
Yeah I think that is correct. I really need some better tests for this resampling business at some point (cough Pole Figures cough). I've also added volume fractions in the return, https://github.com/Patol75/PyDRex/commit/b03ba3742286f68fbb662c5b1ff6c5792011ba3d#diff-e6b4c1fadcae25149a1eb99ee1934f02ef26094e179e218dc2116af77edce5f7R102-R108